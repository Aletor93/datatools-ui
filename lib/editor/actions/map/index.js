import ll from '@conveyal/lonlat'
import clone from 'lodash.clonedeep'
import uniqueId from 'lodash.uniqueid'
import {createAction} from 'redux-actions'
import shp from 'shpjs'
import lineDistance from 'turf-line-distance'
import lineString from 'turf-linestring'
import nearestPointOnLine from '@turf/nearest-point-on-line'
import point from 'turf-point'

import {POINT_TYPE} from '../../constants'
import {setActivePatternSegment} from '../tripPattern'
import {setErrorMessage} from '../../../manager/actions/status'
import {getLineSlices, recalculateShape, newControlPoint} from '../../util/map'

// Simple actions
export const updatePatternGeometry = createAction('UPDATE_PATTERN_GEOMETRY')
const controlPointDragOrEnd = createAction('CONTROL_POINT_DRAG_START_OR_END')
const receivedRoutesShapefile = createAction('RECEIVED_ROUTES_SHAPEFILE')
// Update map setting is used in complex actions and elsewhere (editor components).
export const updateMapSetting = createAction('UPDATE_MAP_SETTING')

// FIXME: replace use of pattern.shape with shape points
export function constructControlPoint ({pattern, latlng, controlPoints, segmentIndex, segmentCoordinates, patternCoordinates}) {
  return function (dispatch, getState) {
    const {patternSegment} = getState().editor.data.active
    // console.log(patternSegment, segmentIndex)
    const patternLine = lineString(segmentCoordinates)
    const coord = ll.toCoordinates(latlng)
    const clickPoint = point(coord)
    // Assign point to location on line (click latlng may not be precise at certain zoom levels)
    const pointOnLine = nearestPointOnLine(patternLine, clickPoint)
    // slice line
    const {beforeSlice, afterSlice} = getLineSlices(patternLine, pointOnLine)

    // measure line segment
    const distTraveled = lineDistance(beforeSlice, 'meters')
    // FIXME: Check that shape pt sequence value here is accurate
    const controlPoint = newControlPoint(distTraveled, pointOnLine, {
      shapePtSequence: beforeSlice.geometry.coordinates.length,
      shapePtLat: pointOnLine.geometry.coordinates[1],
      shapePtLon: pointOnLine.geometry.coordinates[0],
      // User can only add control points for non-stop anchors
      pointType: POINT_TYPE.ANCHOR // 1
    })

    // console.log(controlPoints, controlPoint)

    // Make a copy of the latest set of control points
    const updatedControlPoints = [...controlPoints]
    // Splice in new control point.
    const controlPointIndex = segmentIndex + 1
    updatedControlPoints.splice(controlPointIndex, 0, controlPoint)
    const patternSegments = clone(patternCoordinates)
    // Update item at segmentIndex and insert new segment
    patternSegments.splice(segmentIndex, 1, beforeSlice.geometry.coordinates, afterSlice.geometry.coordinates)
    dispatch(updatePatternGeometry({
      controlPoints: updatedControlPoints,
      patternSegments
    }))
    dispatch(setActivePatternSegment(controlPointIndex))
  }
}

export function displayRoutesShapefile (feedSource, file) {
  return function (dispatch, getState) {
    const reader = new window.FileReader()
    reader.onload = e => {
      const arrayBuff = reader.result
      shp(arrayBuff).then(geojson => {
        console.log(geojson)
        geojson.key = Math.random()
        dispatch(receivedRoutesShapefile({feedSource, geojson}))
      })
    }
    reader.readAsArrayBuffer(file)
  }
}

/**
 * Calculate a new shape according to newly dragged position of control point
 *
 * @param  {ControlPoint[]} controlPoints
 * @param  {Number} index Index of the control point being dragged
 * @param  {Position} latlng  Position of drag point
 * @param  {Object} patternShape
 */
export function handleControlPointDrag (
  controlPoints,
  index,
  latlng,
  pattern,
  patternCoordinates
) {
  return function (dispatch, getState) {
    const {currentDragId, followStreets} = getState().editor.editSettings.present
    recalculateShape({
      controlPoints,
      defaultToStraightLine: false,
      dragId: currentDragId,
      editType: 'update',
      followStreets,
      index,
      newPoint: latlng,
      pattern,
      patternCoordinates
    }).then(result => {
      const {currentDragId} = getState().editor.editSettings.present
      // If there is a dragId in the store and it matches the result, the user
      // is still dragging the control point
      const stillDragging = currentDragId && result.dragId === currentDragId
      if (
        stillDragging &&
        result.coordinates
      ) {
        // console.log('drag result', result.coordinates)
        dispatch(updatePatternGeometry({
          controlPoints: result.updatedControlPoints,
          patternSegments: result.coordinates
        }))
      }
    })
  }
}

export function handleControlPointDragEnd (controlPoints, index, latlng, pattern, patternCoordinates) {
  return function (dispatch, getState) {
    // proclaim end of dragging (with undefined dragId argument)
    dispatch(controlPointDragOrEnd())

    // recalculate shape for final position
    const {followStreets} = getState().editor.editSettings.present
    recalculateShape({
      controlPoints,
      defaultToStraightLine: false,
      editType: 'update',
      index,
      followStreets,
      newPoint: latlng,
      pattern,
      patternCoordinates,
      snapControlPointToNewSegment: true
    }).then(result => {
      // const {updatedShapePoints: shapePoints, updatedControlPoints} = result
      if (!result.coordinates) {
        dispatch(setErrorMessage({
          message: 'An error occurred while trying to recalculate the shape after dragging.  Please try again.'
        }))
      } else {
        // console.log('drag end result', result)
        dispatch(updatePatternGeometry({
          controlPoints: result.updatedControlPoints,
          patternSegments: result.coordinates
        }))
      }
    })
  }
}

/**
 * Save data to store that dragging is in progress
 *
 * @param  {ControlPoint} controlPoint
 * @return {Action}
 */
export function handleControlPointDragStart (controlPoint) {
  return controlPointDragOrEnd(uniqueId(controlPoint.id))
}

export function removeControlPoint (controlPoints, index, pattern, patternCoordinates) {
  return async function (dispatch, getState) {
    const {followStreets} = getState().editor.editSettings.present
    const {
      coordinates,
      updatedControlPoints
    } = await recalculateShape({
      controlPoints,
      editType: 'delete',
      index,
      followStreets,
      pattern,
      patternCoordinates
    })
    // Update active pattern in store (does not save to server).
    dispatch(updatePatternGeometry({
      controlPoints: updatedControlPoints,
      patternSegments: coordinates
    }))
  }
}
