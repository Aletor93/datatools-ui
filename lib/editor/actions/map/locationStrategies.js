// @flow

import type { dispatchFn, getStateFn } from '../../../types/reducers'
import { updatingLocationShape } from '../active'
import type { GeometryType, LocationShape } from '../../../types'

/**
 * Updates the location shape (Polygon, polyline, multi polygon or polyline)
 */
export function updateLocationShape (
  geometryType: GeometryType,
  shapes: LocationShape
) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    return dispatch(updatingLocationShape({geometryType, shapes}))
  }
}
