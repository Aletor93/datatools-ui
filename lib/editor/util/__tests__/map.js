/* globals describe, expect, it */

import nock from 'nock'

import { recalculateShape } from '../map'

const fixtureControlPoints = require('./fixtures/test-control-points.json')
const followStreets = true
const patternCoordinates = require('./fixtures/test-pattern-shape.json')

describe('editor > util > map >', () => {
  describe('recalculateShape >', () => {
    function makeTestCase ({
      controlPoints = fixtureControlPoints,
      editType,
      index,
      graphHopperResponseFile,
      newPoint,
      resultCoordsCanBeNull = false,
      title
    }) {
      return it(
        title,
        async () => {
          // setup nock
          if (graphHopperResponseFile) {
            nock('https://graphhopper.com')
              // Turn on log in order to view nock request URLs.
              // .log(console.log)
              .get(/\/api\/1\/*/)
              .reply(200, require(graphHopperResponseFile))
          }

          const result = await recalculateShape({
            controlPoints,
            editType,
            index,
            followStreets,
            newPoint,
            patternCoordinates
          })

          expect(result).toMatchSnapshot()
          // Coordinates is an array of segments, each of which is an array of
          // coordinate pairs.
          const { coordinates } = result
          if (!resultCoordsCanBeNull) {
            expect(coordinates.length).toBeGreaterThan(0)
            coordinates.forEach(segment => {
              segment.forEach(coord => {
                // Expect each item to be a coordinate pair (lon, lat).
                expect(coord.length).toBe(2)
                expect(Math.abs(coord[0])).toBeLessThanOrEqual(180)
                expect(Math.abs(coord[1])).toBeLessThanOrEqual(90)
              })
            })
          }
        }
      )
    }

    describe('updating >', () => {
      makeTestCase({
        editType: 'update',
        index: 0,
        graphHopperResponseFile: './fixtures/graphhopper-response-update-first-stop.json',
        newPoint: [-77.92553, 42.08666],
        title: 'should update the first stop'
      })

      makeTestCase({
        editType: 'update',
        index: 4,
        graphHopperResponseFile: './fixtures/graphhopper-response-update-last-stop.json',
        newPoint: [-77.92262, 42.09828],
        title: 'should update the last stop'
      })

      makeTestCase({
        editType: 'update',
        index: 3,
        graphHopperResponseFile: './fixtures/graphhopper-response-update-middle-control-point.json',
        newPoint: [-77.92382, 42.09182],
        title: 'should update a point in the middle of the control points'
      })
    })

    describe('deleting >', () => {
      makeTestCase({
        editType: 'delete',
        index: 0,
        title: 'should delete the first stop'
      })

      makeTestCase({
        editType: 'delete',
        index: 4,
        title: 'should delete the last stop'
      })

      makeTestCase({
        controlPoints: require('./fixtures/test-control-points-with-extra-point-at-end.json'),
        editType: 'delete',
        index: 2,
        resultCoordsCanBeNull: true,
        title: 'should delete the last stop when there are control points after the last stop'
      })

      makeTestCase({
        editType: 'delete',
        index: 3,
        graphHopperResponseFile: './fixtures/graphhopper-response-delete-middle-control-point.json',
        title: 'should delete a point in the middle of the control points'
      })
    })
  })
})
