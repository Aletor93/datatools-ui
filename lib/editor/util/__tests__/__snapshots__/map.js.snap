// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor > util > map > recalculateShape > deleting > should delete a point in the middle of the control points 1`] = `
Object {
  "coordinates": Array [
    Array [
      Array [
        -77.922203,
        42.086407,
      ],
      Array [
        -77.922196,
        42.086498,
      ],
      Array [
        -77.922173,
        42.086498,
      ],
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
    ],
    Array [
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
      Array [
        -77.92188300000001,
        42.086956,
      ],
      Array [
        -77.921929,
        42.087368,
      ],
      Array [
        -77.922043,
        42.088386,
      ],
      Array [
        -77.922112,
        42.088844,
      ],
      Array [
        -77.922135,
        42.089099,
      ],
    ],
    Array [
      Array [
        -77.92252574804179,
        42.09235127309673,
      ],
      Array [
        -77.92253,
        42.09235,
      ],
      Array [
        -77.92269,
        42.09374,
      ],
      Array [
        -77.92306,
        42.09616,
      ],
    ],
  ],
  "dragId": undefined,
  "updatedControlPoints": Array [
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 0,
      "distance": 0,
      "hidden": true,
      "id": "bv67",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.922203,
            42.086407,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 0,
      "stopId": "56c69e57-b007-4367-94ed-2c224662a2af",
      "timepoint": null,
    },
    Object {
      "distance": 345.8645390523855,
      "id": "y1e5",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92215494185281,
            42.08925407008939,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 120,
      "distance": 691.729078104771,
      "hidden": true,
      "id": "fyz2",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92252574804179,
            42.09235127309673,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 691.729078104771,
      "stopId": "a9c9cd55-ca3f-4ce7-9a93-96f2f4eaf089",
      "timepoint": null,
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 240,
      "distance": 1118.1829618006304,
      "hidden": true,
      "id": "1gwo",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92305462145765,
            42.096168018547736,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 1118.1829618006304,
      "shapePtSequence": NaN,
      "stopId": "5b6afe4d-f4f2-43f3-810e-91749420af53",
      "timepoint": null,
    },
  ],
}
`;

exports[`editor > util > map > recalculateShape > deleting > should delete the first stop 1`] = `
Object {
  "coordinates": Array [
    Array [
      Array [
        -77.922135,
        42.089099,
      ],
      Array [
        -77.92231799999999,
        42.090522,
      ],
      Array [
        -77.92235600000001,
        42.090847000000004,
      ],
      Array [
        -77.922432,
        42.091491000000005,
      ],
      Array [
        -77.922509,
        42.092201,
      ],
      Array [
        -77.922654,
        42.093502,
      ],
      Array [
        -77.922684,
        42.093745999999996,
      ],
    ],
    Array [
      Array [
        -77.922684,
        42.093745999999996,
      ],
      Array [
        -77.922707,
        42.09391,
      ],
      Array [
        -77.92274499999999,
        42.094158,
      ],
      Array [
        -77.922799,
        42.094528000000004,
      ],
      Array [
        -77.922974,
        42.095657,
      ],
      Array [
        -77.923104,
        42.096481,
      ],
    ],
  ],
  "dragId": undefined,
  "updatedControlPoints": Array [
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 120,
      "distance": 0,
      "hidden": true,
      "id": "fyz2",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92252574804179,
            42.09235127309673,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 0,
      "shapePtSequence": NaN,
      "stopId": "a9c9cd55-ca3f-4ce7-9a93-96f2f4eaf089",
      "timepoint": null,
    },
    Object {
      "distance": 213.4093278296342,
      "id": "0cj3",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92276015989871,
            42.09426187381858,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": NaN,
      "shapePtSequence": NaN,
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 240,
      "distance": 426.81865565926853,
      "hidden": true,
      "id": "1gwo",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92305462145765,
            42.096168018547736,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 426.81865565926853,
      "shapePtSequence": NaN,
      "stopId": "5b6afe4d-f4f2-43f3-810e-91749420af53",
      "timepoint": null,
    },
  ],
}
`;

exports[`editor > util > map > recalculateShape > deleting > should delete the last stop 1`] = `
Object {
  "coordinates": Array [
    Array [
      Array [
        -77.922203,
        42.086407,
      ],
      Array [
        -77.922196,
        42.086498,
      ],
      Array [
        -77.922173,
        42.086498,
      ],
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
    ],
    Array [
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
      Array [
        -77.92188300000001,
        42.086956,
      ],
      Array [
        -77.921929,
        42.087368,
      ],
      Array [
        -77.922043,
        42.088386,
      ],
      Array [
        -77.922112,
        42.088844,
      ],
      Array [
        -77.922135,
        42.089099,
      ],
    ],
  ],
  "dragId": undefined,
  "updatedControlPoints": Array [
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 0,
      "distance": 0,
      "hidden": true,
      "id": "bv67",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.922203,
            42.086407,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 0,
      "stopId": "56c69e57-b007-4367-94ed-2c224662a2af",
      "timepoint": null,
    },
    Object {
      "distance": 345.8645390523855,
      "id": "y1e5",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92215494185281,
            42.08925407008939,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 120,
      "distance": 691.729078104771,
      "hidden": true,
      "id": "fyz2",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92252574804179,
            42.09235127309673,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 691.729078104771,
      "stopId": "a9c9cd55-ca3f-4ce7-9a93-96f2f4eaf089",
      "timepoint": null,
    },
  ],
}
`;

exports[`editor > util > map > recalculateShape > deleting > should delete the last stop when there are control points after the last stop 1`] = `
Object {
  "coordinates": Array [
    Array [
      Array [
        -77.922203,
        42.086407,
      ],
      Array [
        -77.922196,
        42.086498,
      ],
      Array [
        -77.922173,
        42.086498,
      ],
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
    ],
    Array [
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
      Array [
        -77.92188300000001,
        42.086956,
      ],
      Array [
        -77.921929,
        42.087368,
      ],
      Array [
        -77.922043,
        42.088386,
      ],
      Array [
        -77.922112,
        42.088844,
      ],
      Array [
        -77.922135,
        42.089099,
      ],
    ],
    Array [
      Array [
        -77.922135,
        42.089099,
      ],
      Array [
        -77.92231799999999,
        42.090522,
      ],
      Array [
        -77.92235600000001,
        42.090847000000004,
      ],
      Array [
        -77.922432,
        42.091491000000005,
      ],
      Array [
        -77.922509,
        42.092201,
      ],
      Array [
        -77.922654,
        42.093502,
      ],
      Array [
        -77.922684,
        42.093745999999996,
      ],
    ],
    Array [
      Array [
        -77.922684,
        42.093745999999996,
      ],
      Array [
        -77.922707,
        42.09391,
      ],
      Array [
        -77.92274499999999,
        42.094158,
      ],
      Array [
        -77.922799,
        42.094528000000004,
      ],
      Array [
        -77.922974,
        42.095657,
      ],
      Array [
        -77.923104,
        42.096481,
      ],
    ],
  ],
  "updatedControlPoints": Array [
    Object {
      "distance": 0,
      "id": "2zu5",
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -122.00743318,
            37.06121322,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "pointType": 2,
      "shapePtSequence": 0,
      "stopId": "aad0f374-0f29-4300-ae05-6affbe0f1562",
    },
    Object {
      "distance": 170.7245980700509,
      "id": "25p4",
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -122.00872656844231,
            37.060454715231224,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "pointType": 1,
      "shapePtSequence": 1,
    },
    Object {
      "distance": 341.4491961401018,
      "id": "cy46",
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -122.00981702504365,
            37.059191439360475,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "pointType": 1,
      "shapePtSequence": 2,
      "stopId": undefined,
    },
    Object {
      "distance": 358.0806866950734,
      "id": "4s05",
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -122.009919,
            37.059066,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "pointType": 1,
      "shapePtSequence": 3,
    },
  ],
}
`;

exports[`editor > util > map > recalculateShape > updating > should update a point in the middle of the control points 1`] = `
Object {
  "coordinates": Array [
    Array [
      Array [
        -77.922203,
        42.086407,
      ],
      Array [
        -77.922196,
        42.086498,
      ],
      Array [
        -77.922173,
        42.086498,
      ],
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
    ],
    Array [
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
      Array [
        -77.92188300000001,
        42.086956,
      ],
      Array [
        -77.921929,
        42.087368,
      ],
      Array [
        -77.922043,
        42.088386,
      ],
      Array [
        -77.922112,
        42.088844,
      ],
      Array [
        -77.922135,
        42.089099,
      ],
    ],
    Array [
      Array [
        -77.92252574804179,
        42.09235127309673,
      ],
      Array [
        -77.92253,
        42.09235,
      ],
      Array [
        -77.92251,
        42.0922,
      ],
      Array [
        -77.92264,
        42.09218,
      ],
      Array [
        -77.92374,
        42.09218,
      ],
      Array [
        -77.92374975804022,
        42.091818952512064,
      ],
    ],
    Array [
      Array [
        -77.92374975804022,
        42.091818952512064,
      ],
      Array [
        -77.92375,
        42.09181,
      ],
      Array [
        -77.92374,
        42.09218,
      ],
      Array [
        -77.92264,
        42.09218,
      ],
      Array [
        -77.92251,
        42.0922,
      ],
      Array [
        -77.92269,
        42.09374,
      ],
      Array [
        -77.92306,
        42.09616,
      ],
    ],
  ],
  "dragId": undefined,
  "updatedControlPoints": Array [
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 0,
      "distance": 0,
      "hidden": true,
      "id": "bv67",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.922203,
            42.086407,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 0,
      "stopId": "56c69e57-b007-4367-94ed-2c224662a2af",
      "timepoint": null,
    },
    Object {
      "distance": 345.8645390523855,
      "id": "y1e5",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92215494185281,
            42.08925407008939,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 120,
      "distance": 691.729078104771,
      "hidden": true,
      "id": "fyz2",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92252574804179,
            42.09235127309673,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 691.729078104771,
      "stopId": "a9c9cd55-ca3f-4ce7-9a93-96f2f4eaf089",
      "timepoint": null,
    },
    Object {
      "distance": 905.1384059344052,
      "id": "0cj3",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92382,
            42.09182,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 240,
      "distance": 1437.544930172152,
      "hidden": true,
      "id": "1gwo",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92305462145765,
            42.096168018547736,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 1437.544930172152,
      "shapePtSequence": NaN,
      "stopId": "5b6afe4d-f4f2-43f3-810e-91749420af53",
      "timepoint": null,
    },
  ],
}
`;

exports[`editor > util > map > recalculateShape > updating > should update the first stop 1`] = `
Object {
  "coordinates": Array [
    Array [
      Array [
        -77.92553,
        42.08666,
      ],
      Array [
        -77.92553,
        42.0867,
      ],
      Array [
        -77.92369,
        42.0867,
      ],
      Array [
        -77.92332,
        42.08668,
      ],
      Array [
        -77.9232,
        42.08666,
      ],
      Array [
        -77.92292,
        42.08654,
      ],
      Array [
        -77.92286,
        42.08652,
      ],
      Array [
        -77.92217,
        42.08649,
      ],
      Array [
        -77.92184,
        42.08651,
      ],
      Array [
        -77.92216,
        42.08925,
      ],
    ],
    Array [
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
      Array [
        -77.92188300000001,
        42.086956,
      ],
      Array [
        -77.921929,
        42.087368,
      ],
      Array [
        -77.922043,
        42.088386,
      ],
      Array [
        -77.922112,
        42.088844,
      ],
      Array [
        -77.922135,
        42.089099,
      ],
    ],
    Array [
      Array [
        -77.922135,
        42.089099,
      ],
      Array [
        -77.92231799999999,
        42.090522,
      ],
      Array [
        -77.92235600000001,
        42.090847000000004,
      ],
      Array [
        -77.922432,
        42.091491000000005,
      ],
      Array [
        -77.922509,
        42.092201,
      ],
      Array [
        -77.922654,
        42.093502,
      ],
      Array [
        -77.922684,
        42.093745999999996,
      ],
    ],
    Array [
      Array [
        -77.922684,
        42.093745999999996,
      ],
      Array [
        -77.922707,
        42.09391,
      ],
      Array [
        -77.92274499999999,
        42.094158,
      ],
      Array [
        -77.922799,
        42.094528000000004,
      ],
      Array [
        -77.922974,
        42.095657,
      ],
      Array [
        -77.923104,
        42.096481,
      ],
    ],
  ],
  "dragId": undefined,
  "updatedControlPoints": Array [
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 0,
      "distance": 0,
      "hidden": true,
      "id": "bv67",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92553,
            42.08666,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 0,
      "stopId": "56c69e57-b007-4367-94ed-2c224662a2af",
      "timepoint": null,
    },
    Object {
      "distance": 619.530652509,
      "id": "y1e5",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92215494185281,
            42.08925407008939,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": NaN,
      "shapePtSequence": NaN,
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 120,
      "distance": 965.3951915613854,
      "hidden": true,
      "id": "fyz2",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92252574804179,
            42.09235127309673,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 965.3951915613854,
      "shapePtSequence": NaN,
      "stopId": "a9c9cd55-ca3f-4ce7-9a93-96f2f4eaf089",
      "timepoint": null,
    },
    Object {
      "distance": 1178.8045193910198,
      "id": "0cj3",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92276015989871,
            42.09426187381858,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": NaN,
      "shapePtSequence": NaN,
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 240,
      "distance": 1392.213847220654,
      "hidden": true,
      "id": "1gwo",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92305462145765,
            42.096168018547736,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 1392.213847220654,
      "shapePtSequence": NaN,
      "stopId": "5b6afe4d-f4f2-43f3-810e-91749420af53",
      "timepoint": null,
    },
  ],
}
`;

exports[`editor > util > map > recalculateShape > updating > should update the last stop 1`] = `
Object {
  "coordinates": Array [
    Array [
      Array [
        -77.922203,
        42.086407,
      ],
      Array [
        -77.922196,
        42.086498,
      ],
      Array [
        -77.922173,
        42.086498,
      ],
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
    ],
    Array [
      Array [
        -77.92183700000001,
        42.086513000000004,
      ],
      Array [
        -77.92188300000001,
        42.086956,
      ],
      Array [
        -77.921929,
        42.087368,
      ],
      Array [
        -77.922043,
        42.088386,
      ],
      Array [
        -77.922112,
        42.088844,
      ],
      Array [
        -77.922135,
        42.089099,
      ],
    ],
    Array [
      Array [
        -77.922135,
        42.089099,
      ],
      Array [
        -77.92231799999999,
        42.090522,
      ],
      Array [
        -77.92235600000001,
        42.090847000000004,
      ],
      Array [
        -77.922432,
        42.091491000000005,
      ],
      Array [
        -77.922509,
        42.092201,
      ],
      Array [
        -77.922654,
        42.093502,
      ],
      Array [
        -77.922684,
        42.093745999999996,
      ],
    ],
    Array [
      Array [
        -77.92276015989871,
        42.09426187381858,
      ],
      Array [
        -77.92276,
        42.09426,
      ],
      Array [
        -77.92311,
        42.09647,
      ],
      Array [
        -77.92323,
        42.09698,
      ],
      Array [
        -77.92335,
        42.09735,
      ],
      Array [
        -77.92364,
        42.09798,
      ],
    ],
    Array [
      Array [
        -77.92364,
        42.09798,
      ],
      Array [
        -77.92364,
        42.09798,
      ],
    ],
  ],
  "dragId": undefined,
  "updatedControlPoints": Array [
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 0,
      "distance": 0,
      "hidden": true,
      "id": "bv67",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.922203,
            42.086407,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 0,
      "stopId": "56c69e57-b007-4367-94ed-2c224662a2af",
      "timepoint": null,
    },
    Object {
      "distance": 345.8645390523855,
      "id": "y1e5",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92215494185281,
            42.08925407008939,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 120,
      "distance": 691.729078104771,
      "hidden": true,
      "id": "fyz2",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92252574804179,
            42.09235127309673,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 691.729078104771,
      "stopId": "a9c9cd55-ca3f-4ce7-9a93-96f2f4eaf089",
      "timepoint": null,
    },
    Object {
      "distance": 905.1384059344052,
      "id": "0cj3",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92276015989871,
            42.09426187381858,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
    },
    Object {
      "defaultDwellTime": 0,
      "defaultTravelTime": 240,
      "distance": 1118.5477337640395,
      "hidden": true,
      "id": "1gwo",
      "permanent": true,
      "point": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.92262,
            42.09828,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      "shapeDistTraveled": 1118.5477337640395,
      "stopId": "5b6afe4d-f4f2-43f3-810e-91749420af53",
      "timepoint": null,
    },
  ],
}
`;
